keps/sig-storage/1702-volume-provisioning-rollback/README.md

```diff
@@ -0,0 +1,346 @@
+issue in [kubernetes/enhancements] referencing this KEP and targeting a release
+milestone **before the [Enhancement Freeze](https://git.k8s.io/sig-release/releases)
+of the targeted release**.
+
+For enhancements that make changes to code or processes/procedures in core
+Kubernetes i.e., [kubernetes/kubernetes], we require the following Release
+Signoff checklist to be completed.
+
+Check these off as they are completed for the Release Team to track. These
+checklist items _must_ be updated for the enhancement to be released.
+-->
+
+- [ ] Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] KEP approvers have approved the KEP status as `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+When a pod uses one or more volumes with late binding (aka
+`WaitForFirstConsumer`), volume provisioning is kicked of when a pod
+gets tentatively assigned to a node. When one of the volumes cannot be
+provisioned for the selected node, the scheduler needs to try with
+some other node. But it cannot do that if the already provisioned
+volumes are only available on the original node.
+
+The KEP proposes a rollback mechanism that undos the provisioning of
+volumes that block the rescheduling of a pod.
+
+## Motivation
+
+Storage capacity aware pod scheduling increases the chance that a node
+is picked which still has enough capacity left. However, there are
+still cases (race conditions, insufficient information) where the
+actual attempts to create volumes fail.
+
+If scheduling was based on stale information, the next attempt may be
+able to pick a better node. However for that to work, Kubernetes has
+to prevent getting stuck with partially allocated resources for the
+pod.
+
+### Goals
+
+- Enable recovery from a bad node selection for a pod when volume
+  provisioning fails for some, but not all nodes.
+
+### Non-Goals
+
+- Only CSI drivers using an up-to-date external-provisioner will be
+  supported.
+
+## Proposal
+
+The proposal is to mark provisioned PVCs as "unused" via a label until
```

Comment:
Who sets this label to "used"? How do we ensure there aren't races between the time we check the label and when we go to delete the PV?

Reply:
Perhaps we can solve this entirely in the external-provisioner: in addition to the "selected node" label that it gets already now, it also needs to know which pod(s) are triggering the provisioning so that it can group several PVCs together.

Then we need an intermediate phase where external-provisioner creates PVs, but doesn't finally bind them to the PVCs yet. That's the "unused" state. Then when all volumes have been created, external-provisioner finishes the binding.

But it will still be tricky to update several objects together atomically :cry:

Reply:
Scratch that: there is more than one external-provisioner involved, so this is the wrong place.

A better place is the volume scheduling code. That's where kube-scheduler checks whether a pod can move ahead; we can block that until all volumes have been provisioned. Before giving it's okay, that code can also remove the "unused" flag.